### ISSUE TITLE
nanoT5 for different embeddings

### ISSUE THREAD

**Issue by victoriazinkovich** [PiotrNawrot/nanoT5#40]
Dead Piotr, 
Thank you for such a wonderful repository! Could you please give me some advice on the problem below? I have been struggling to find an answer for over a month now... :'(

I am trying to train the nanoT5 model so that it would be capable to take LISA frozen embeddings with its own tokenizer. I have changed nn.Embedding layer to nn.CustomEmbedding (such that it would take embeddings of size 5120), changed tokenizer and some special tokens. The **interesting thing** is that the model perfectly passes the pre-training stage with accuracy of 63.3%. However, when doing fine-tuning, the model performance is very poor, with Rouge-L approximately equals 3.5. Do you maybe know what am I doing wrong? :( 
I would be extremely grateful for any feedback! 

Here is my code and graphics: 

Pre-training:
![image](https://github.com/user-attachments/assets/2b8838ff-a05d-4a7f-b6c3-dbd8926b2368)

Fine-tuning:
![image](https://github.com/user-attachments/assets/0b987ef4-abe5-4cf4-b772-23efee2df94b)
 
Code:
`t5_model.py`
```
class CustomEmbedding(nn.Module):
    def __init__(self, target_dim, model_name):
        super().__init__()
        self.lisa_model = LISAForCausalLM.from_pretrained(model_name, low_cpu_mem_usage=True)
        lisa_embeddings = self.lisa_model.get_input_embeddings().weight.clone().detach()
        self.embedding = nn.Embedding.from_pretrained(lisa_embeddings, freeze=True)
        self.adapter = nn.Sequential(
            nn.Linear(lisa_embeddings.shape[1], lisa_embeddings.shape[1]),
            nn.GELU(),
            nn.Linear(lisa_embeddings.shape[1], target_dim),
        )

    def forward(self, input_ids):
        embeddings = self.embedding(input_ids)
        return self.adapter(embeddings)
```
`model_utils.py`
```
def get_tokenizer(args):
    tokenizer = AutoTokenizer.from_pretrained(
        "xinlai/LISA-13B-llama2-v1",
        use_fast=False
    )
    tokenizer.pad_token = tokenizer.unk_token
    tokenizer.model_max_length = int(1e9)
    return tokenizer
```
`main.py`
```
tokenizer = get_tokenizer(args)
config.eos_token_id = tokenizer.eos_token_id
config.bos_token_id = tokenizer.bos_token_id
config.decoder_start_token_id = tokenizer.bos_token_id
config.pad_token_id = tokenizer.pad_token_id
config.vocab_size = len(tokenizer.get_vocab())
model = get_model(args, config)
```

Parameters in configs are original. I have tried to search for different hyperparameters, but results stay the same.

**Comment by PiotrNawrot** [comment]
That's in fact very weird.

First of all, could you try fine-tuning phase with randomly initialised T5? I have provided the loss curves [here](https://github.com/PiotrNawrot/nanoT5/blob/main/assets/ft_loss.png). This run ended up with ~10 RougeL, even though the training loss was ~4.5 and test loss was ~7.5, so your low RougeL values really surprise me. I want you to try randomly initialised T5 and maybe try fine-tuning the T5 model I put on HF Hub, just to make sure that your fine-tuning / evaluation setup is alright.

One thing that seems a bit weird to me is that compared to my fine-tuning curves, yours are way worse. You start from a very high spot in terms of perplexity which maybe suggests that you are loading the weights after pre-training wrongly? Maybe you reinitialize something from scratch, maybe you overwrite something? Could you try loading the model the way you would load it for fine-tuning, but instead evaluate accuracy and perplexity on the pre-training dataset - this would be to confirm that the values you get at the end of the training are the same and that you're loading the model correctly.

As you said, the curves for pre-training look alright, but there seems to be something wrong with fine-tuning. I'd recommend reproducing my results for fine-tuning, and then make sure that you're loading the model correctly after pre-training phase. 

Let me know if that will work! Good luck. Feel free to re-open in case you have more questions.

**Comment by victoriazinkovich** [comment]
> That's in fact very weird.
> 
> First of all, could you try fine-tuning phase with randomly initialised T5? I have provided the loss curves [here](https://github.com/PiotrNawrot/nanoT5/blob/main/assets/ft_loss.png). This run ended up with ~10 RougeL, even though the training loss was ~4.5 and test loss was ~7.5, so your low RougeL values really surprise me. I want you to try randomly initialised T5 and maybe try fine-tuning the T5 model I put on HF Hub, just to make sure that your fine-tuning / evaluation setup is alright.
> 
> One thing that seems a bit weird to me is that compared to my fine-tuning curves, yours are way worse. You start from a very high spot in terms of perplexity which maybe suggests that you are loading the weights after pre-training wrongly? Maybe you reinitialize something from scratch, maybe you overwrite something? Could you try loading the model the way you would load it for fine-tuning, but instead evaluate accuracy and perplexity on the pre-training dataset - this would be to confirm that the values you get at the end of the training are the same and that you're loading the model correctly.
> 
> As you said, the curves for pre-training look alright, but there seems to be something wrong with fine-tuning. I'd recommend reproducing my results for fine-tuning, and then make sure that you're loading the model correctly after pre-training phase.
> 
> Let me know if that will work! Good luck. Feel free to re-open in case you have more questions.

Thank you very much for your prompt and detailed advice! I really appreciate it! :))
